app/src/main/java/com/realityexpander/tasky/core/presentation/notifications/RemindAtAlarmManager.kt

```diff
@@ -0,0 +1,185 @@
+import com.realityexpander.tasky.agenda_feature.domain.AgendaItem
+import com.realityexpander.tasky.agenda_feature.presentation.common.enums.toAgendaItemType
+import com.realityexpander.tasky.core.util.toUtcMillis
+import com.realityexpander.tasky.core.util.toZonedDateTime
+import logcat.logcat
+import java.time.ZonedDateTime
+import com.realityexpander.tasky.core.presentation.notifications.RemindAtAlarmNotificationManager as Notifications
+
+interface IRemindAtAlarmManager {
+
+    fun setAlarmsForAgendaItems(
+        context: Context,
+        agendaItems: List<AgendaItem>
+    )
+
+    fun cancelAllAlarms(
+        context: Context,
+        onFinished: () -> Unit = {},
+    )
+}
+
+object RemindAtAlarmManager : IRemindAtAlarmManager {
+
+    private const val CURRENT_ALARM_PENDING_INTENTS = "CURRENT_ALARM_PENDING_INTENTS"
+    private const val CURRENT_ALARM_TITLES = "CURRENT_ALARM_TITLES"
+
+    private const val REMIND_AT_ALARM_SUPERVISOR_REQUEST_CODE = 0
+
+    override fun setAlarmsForAgendaItems(
+        context: Context,
+        agendaItems: List<AgendaItem>
+    ) {
+        // Only include upcoming `Remind At` items (RemindAt is in the future)
+        val futureItems =
+            agendaItems.filter {
+                it.remindAtTime.toUtcMillis() >= ZonedDateTime.now().toUtcMillis()
+            }
+
+        if (futureItems.isEmpty()) {
+            return
+        }
+
+        // Create an Alarm PendingIntent for each AgendaItem's RemindAt time
+        val alarmPendingIntents = mutableListOf<PendingIntent>()
+        futureItems.forEachIndexed { alarmIndex, agendaItem ->
+            val pendingIntent =
+                createAlarmPendingIntentForAgendaItem(context, agendaItem, alarmIndex + 1)
+            pendingIntent.let {
+                alarmPendingIntents.add(it)
+            }
+        }
+
+        // Save all current Alarm PendingIntents to an "Alarm Supervisor" PendingIntent
+        //   that will be used to cancel all alarms, when needed.
+        PendingIntent.getBroadcast(
+            context,
+            REMIND_AT_ALARM_SUPERVISOR_REQUEST_CODE,
+            Intent(context, MainActivity::class.java).also {
+                it.putExtra(
+                    CURRENT_ALARM_PENDING_INTENTS,
+                    arrayOf<PendingIntent>(*alarmPendingIntents.toTypedArray())
+                )
+                it.putStringArrayListExtra( // for debugging
+                    CURRENT_ALARM_TITLES,
+                    futureItems.map { agendaItem ->
+                        when (agendaItem) {
+                            is AgendaItem.Event -> agendaItem.title
+                            is AgendaItem.Task -> agendaItem.title
+                            is AgendaItem.Reminder -> agendaItem.title
+                            else -> {
+                                "UNKNOWN_AGENDA_ITEM_TYPE"
+                            }
+                        }
+                    }.toList().toCollection(ArrayList())
```

Comment:
Do you really need to do a .toList() and then .toCollection(ArrayList())?

Reply:
No, i can just to `.toCollection()`

Is there a better way to create an ArrayList from a List?

Reply:
Looks good to me

Reply:
Looks good to me